File: leetcode/linked_list/1836_Remove_Duplicates_From_an_Unsorted_Linked_List/codes/ll.py
from collections import Counter

# Definition for singly-linked list
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def deleteDuplicatesUnsorted(self, head: ListNode) -> ListNode:
        vals = []
        curr = head
        while curr:
            vals.append(curr.val)
            curr = curr.next
        count = Counter(vals)
        dummy = ListNode(0)
        dummy.next = head
        prev = dummy
        curr = head
        while curr:
            if count[curr.val] > 1:
                prev.next = curr.next
            else:
                prev = curr
            curr = curr.next
        return dummy.next

File: leetcode/linked_list/1836_Remove_Duplicates_From_an_Unsorted_Linked_List/codes/ll.cpp
#include <unordered_map>
using namespace std;

/**
Definition for singly-linked list.
 */
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* deleteDuplicatesUnsorted(ListNode* head) {
        unordered_map<int, int> count;
        ListNode* curr = head;
        while (curr) {
            count[curr -> val]++;
            curr = curr -> next;
        }
        ListNode dummy(0);
        dummy.next = head;
        ListNode* prev = &dummy;
        curr = head;
        while (curr) {
            if (count[curr -> val] > 1) {
                prev -> next = curr -> next;
            } else {
                prev = curr;
            }
            curr = curr -> next;
        }
        return dummy.next;
    }
};
